### PR TITLE
Account factory

### PR DESCRIPTION
**Pull request by kress95**
> Previously accepted [here](https://dev.hackerexperience.com/D62)

Add factories for Account and fixes some problems.

### CODE REVIEW

File: apps/account/lib/account/controller/account.ex [hunk 1]
```diff
@@ -4,9 +4,12 @@ defmodule Helix.Account.Controller.Account do
   alias Helix.Account.Model.Account
   alias Helix.Account.Repo
 
+  @type find_params ::
+    [{:email, Account.email}
+    | {:username, Account.username}]
```
Comment (mememori) [inline apps/account/lib/account/controller/account.ex]:
```elixir
@type find_params :: [find_param]
@type find_param :: {:email, Account.email} | {:username, Account.username}
```

File: apps/account/lib/account/controller/account.ex [hunk 2]
```diff
@@ -23,30 +26,33 @@ defmodule Helix.Account.Controller.Account do
...
-  @spec update(Account.id, Account.update_params) :: {:ok, Account}
-    | {:error, Ecto.Changeset.t}
-    | {:error, :notfound}
-  def update(account_id, params) do
-    with {:ok, account} <- find(account_id) do
-      account
-      |> Account.update_changeset(params)
-      |> Repo.update()
-    end
+  @spec reduce_find_params({:email, String.t}, Ecto.Queryable.t) ::
```
Comment (mememori) [inline apps/account/lib/account/controller/account.ex]:
```elixir
@spec reduce_find_params(find_param, Ecto.Queryable.t) :: Ecto.Queryable.t
```

File: apps/account/lib/account/controller/account.ex [hunk 3]
```diff
@@ -23,30 +26,33 @@ defmodule Helix.Account.Controller.Account do
...
+    Ecto.Queryable.t
+  @spec reduce_find_params({:username, String.t}, Ecto.Queryable.t) ::
+    Ecto.Queryable.t
+  defp reduce_find_params({:email, email}, query),
+    do: Account.Query.by_email(query, email)
+  defp reduce_find_params({:username, username}, query),
+    do: Account.Query.by_username(query, username)
+
+  @spec update(Account.t, Account.update_params) ::
+    {:ok, Account} | {:error, Ecto.Changeset.t | :notfound}
```
Comment (mememori) [inline apps/account/lib/account/controller/account.ex]: doesn't return `{:error, :notfound}`

File: apps/account/test/controller/account_test.exs [hunk 4]
```diff
@@ -2,136 +2,155 @@ defmodule Helix.Account.Controller.AccountTest do
...
   end
 
-  describe "find/1" do
-    test "succeeds when account exists", %{account: account} do
+  describe "account fetching" do
+    test "succeeds by id" do
+      account = Factory.insert(:account)
+
       assert {:ok, found} = AccountController.find(account.account_id)
       assert account.account_id == found.account_id
```
Comment (mememori) [inline apps/account/test/controller/account_test.exs]: Remove this line

File: apps/account/test/controller/account_test.exs [hunk 5]
```diff
@@ -2,136 +2,155 @@ defmodule Helix.Account.Controller.AccountTest do
...
     end
 
-    test "fails when account doesn't exists" do
-      assert {:error, :notfound} === AccountController.find(Random.pk())
+    test "succeeds by email" do
+      account = Factory.insert(:account)
+
+      assert [found] = AccountController.find_by(email: account.email)
+      assert account.account_id == found.account_id
```
Comment (mememori) [inline apps/account/test/controller/account_test.exs]: Remove this line

File: apps/account/test/controller/account_test.exs [hunk 6]
```diff
@@ -2,136 +2,155 @@ defmodule Helix.Account.Controller.AccountTest do
...
-  end
 
-  describe "find_by/1" do
-    test "using email", %{account: account} do
-      assert {:ok, found} = AccountController.find_by(email: account.email)
+    test "succeeds by username" do
+      account = Factory.insert(:account)
+
+      assert [found] = AccountController.find_by(username: account.username)
       assert account.account_id == found.account_id
```
Comment (mememori) [inline apps/account/test/controller/account_test.exs]: Remove this line

File: apps/account/lib/account/controller/account.ex (hunk 2 above)
Comment (kress95) [inline apps/account/lib/account/controller/account.ex]: 2 bad that this is getting removed [here](https://github.com/kress95/Helix/tree/fix-account-finds)

File: apps/account/lib/account/controller/account.ex [hunk 7]
```diff
@@ -23,30 +25,30 @@ defmodule Helix.Account.Controller.Account do
...
-      |> Repo.update()
-    end
+  @spec reduce_find_params(find_param, Ecto.Queryable.t) :: Ecto.Queryable.t
+  defp reduce_find_params({:email, email}, query),
+    do: Account.Query.by_email(query, email)
+  defp reduce_find_params({:username, username}, query),
+    do: Account.Query.by_username(query, username)
+
+  @spec update(Account.t, Account.update_params) ::
+    {:ok, Account} | {:error, Ecto.Changeset.t}
```
Comment (kress95) [inline apps/account/lib/account/controller/account.ex]: My bad, it should be okay now.

**Pull request merged**